cloud-devops: begin documenting CD (#1314)
sourcegraph/handbook: content/product-engineering/engineering/cloud/devops/continuous_delivery.md

> # Sourcegraph Cloud continuous delivery

Reviewer: Btw this is a duplicate of this page, we should just update the existing page: https://handbook.sourcegraph.com/product-engineering/engineering/deployments

Reviewer: For things that apply to engineering as a whole, we should try to centralize what we know so that it's easier to maintain and find

Author: Thanks, I wasn't 100% sure where the documentation should live. I'm happy to merge the two documents. Do you think it belongs on [deployments](https://handbook.sourcegraph.com/product-engineering/engineering/deployments) or should I just add further elaboration to [Deploying to sourcegraph.com](https://handbook.sourcegraph.com/product-engineering/engineering/deployments/playbooks#deploying-to-sourcegraphcom)?

Reviewer: If you look at [Deployments basics](https://handbook.sourcegraph.com/product-engineering/engineering/deployments#deployment-basics) it already covers (in a less complete manner) some of the things you've written up here, so I would just update this section

Reviewer: [this also needs to be updated](https://handbook.sourcegraph.com/product-engineering/engineering/deployments#deploy-sourcegraph) to remove references of the renovate downstream/dispatch workflows, those are no longer used

Author: Thank you @bobheadxi . I've made the changes discussed. If you get another chance to take a look at the latest changes that'd be greatly appreciated!

> # Sourcegraph Cloud Continuous Delivery

Reviewer: ```suggestion
# Sourcegraph Cloud continuous delivery
```

IIRC style guide says sentence case

Reviewer: Also I think this is a duplicate of this page: https://handbook.sourcegraph.com/product-engineering/engineering/deployments

This is a brief overview of the CD process for Sourcegraph Cloud (sourcegraph.com).

> 1. New commits pushed to the [mono repo](https://github.com/sourcegraph/sourcegraph) trigger builds by `BuildKite`. The pipelines are all generated [from code](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@main/-/tree/enterprise/dev/ci). To learn more about the build pipelines and the methodology you can read the documentation on [Continuous Integration](https://docs.sourcegraph.com/dev/background-information/continuous_integration).

Reviewer: ```suggestion
1. New commits pushed to the [monorepo](https://github.com/sourcegraph/sourcegraph) trigger builds by `buildkite`. The pipelines are all generated [from code](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@main/-/tree/enterprise/dev/ci). To learn more about the build pipelines and the methodology you can read the documentation on [continuous integration](https://docs.sourcegraph.com/dev/background-information/continuous_integration).
```

>    - Docker images are built by [BuildKite](https://buildkite.com/sourcegraph/sourcegraph) on commit to `main`. These images are pushed to [`GCR`](https://console.cloud.google.com/gcr/images/sourcegraph-dev?project=sourcegraph-dev) with the tag format `<commit-hash>_<build-number>_candidate`.

Reviewer: ```suggestion
   - On `main`, the pipeline will build candidate Docker images that are pushed to [Google Container Registry (internal)](https://console.cloud.google.com/gcr/images/sourcegraph-dev?project=sourcegraph-dev) with the tag format `<commit-hash>_<build-number>_candidate`. The pipeline will also run a suite of tests and checks.
```

>    - Buildkite executes the [`sourcegraph-upgrade`](https://github.com/sourcegraph/sourcegraph/blob/main/.buildkite/vagrant-run.sh) build step in which it spins up a VM to test the build starts correctly.
>    - Buildkite executes [puppeteer tests](https://github.com/sourcegraph/sourcegraph/blob/main/enterprise/dev/ci/internal/ci/operations.go#L178) to test front-end behavior.

Reviewer: ```suggestion
```

This is misleading, it does much more than just that. We should just omit this detail

>    - If both test suites pass the images are "promoted" and pushed to DockerHub with the tag format `<build-number>_<date>_<commit-hash>`.

Reviewer 2: @sourcegraph/dev-experience 
Could you all confirm the above? Or is there a place we should link to? 
@danieldides might be worth adding a line here to say that this is an area of dev-experience "guardianship" 
We are primarily focused on the steps after this. 
Where CI meets CD.

Author: Good call. I just made a comment and linked (I hope) to the dev-ex team on.

Reviewer: ```suggestion
   - If all pipeline steps pass the images are "promoted" and pushed to DockerHub with the tag format `<build-number>_<date>_<commit-hash>`.
```

> 2. [Renovate](https://github.com/sourcegraph/deploy-sourcegraph-dot-com/blob/release/renovate.json5) watches (["after 8am every weekday and before 9am every weekday"](https://github.com/sourcegraph/deploy-sourcegraph-dot-com/blob/release/renovate.json5#L53)) DockerHub for new images. When a new image is found on DockerHub it creates a commit to the [yaml bases](https://github.com/sourcegraph/deploy-sourcegraph-dot-com/tree/release/base) with the new image hash onto the `release` branch.

Reviewer: ```suggestion
2. [Renovate](https://github.com/sourcegraph/deploy-sourcegraph-dot-com/blob/release/renovate.json5) watches ([on a schedule](https://github.com/sourcegraph/deploy-sourcegraph-dot-com/blob/release/renovate.json5#L53)) DockerHub for new images. When a new image is found on DockerHub it creates a commit to the [Kubernetes manifests bases](https://github.com/sourcegraph/deploy-sourcegraph-dot-com/tree/release/base) with the new image tag and image hashes onto the `release` branch.
```

> 3. BuildKite executes an [on-commit pipeline](https://github.com/sourcegraph/deploy-sourcegraph-dot-com/blob/release/.buildkite/pipeline.yaml#L27:L33) to execute [`kubectl-apply-all.sh`](https://github.com/sourcegraph/deploy-sourcegraph-dot-com/blob/release/kubectl-apply-all.sh) with the new image hash and deploy them to the production cluster.

Reviewer: ```suggestion
3. [The `deploy-sourcegraph-dot-com` Buildkite pipeline](https://github.com/sourcegraph/deploy-sourcegraph-dot-com/blob/release/.buildkite/pipeline.yaml#L27:L33) runs [`kubectl-apply-all.sh`](https://github.com/sourcegraph/deploy-sourcegraph-dot-com/blob/release/kubectl-apply-all.sh) with the updated manifests to roll out the new images.
```
